### PR TITLE
PT-3320: Pedigree node hover boxes do not fade out when they should in some cases

### PR DESCRIPTION
**Pull request by allasm**
@sashaandjic please test

### CODE REVIEW

File: components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js [hunk 1]
```diff
@@ -679,11 +679,17 @@ define([
...
                 // only activate "do not hide" check if mouse is still on top of SVG, and not on top of main menu or other windows
+                //
+                // note: some browsers incorrectly report the svgarea's .bottom and .right positions,
+                //       so need to use some adjustment there; the value "3" is derived empirically
                 var svgarea = document.getElementById("work-area").getBoundingClientRect();
-                if (y > svgarea.top && x > svgarea.left && y < svgarea.bottom && x < svgarea.right) {
-                    if (click.x > hoverarea.x && click.x < hoverarea.x2 &&
-                        click.y > hoverarea.y && click.y < hoverarea.y2) {
-                        return;
+                if (y > svgarea.top && x > svgarea.left && y < svgarea.bottom - 3 && x < svgarea.right - 3) {
```
Comment (marta-) [inline components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js]: Not a fan of magic numbers...

File: components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js [hunk 2]
```diff
@@ -679,11 +679,17 @@ define([
...
+                    if (click.x > Math.ceil(hoverarea.x) && click.x < Math.floor(hoverarea.x2) &&
+                        click.y > Math.ceil(hoverarea.y) && click.y < Math.floor(hoverarea.y2)) {
+                        // zoom control behaves in a special way in terms of hoverboxes, so need to treat is specially
+                        if ("relatedTarget" in event && !event.relatedTarget.hasClassName("view-controls-zoom") && !event.relatedTarget.hasClassName("zoom-button")) {
```
Comment (marta-) [inline components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js]: `event.relatedTarget.hasClassName("...")` causes an exception in the console for svg elements, which are nit extended by Prototype.

Comment (marta-) [inline components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js]: Also this doesn't capture the pan controls, and the hoverbox remains open when exiting to the pan controls.

Comment (marta-) [inline components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js]: Exiting through the right and into the legend doesn't hide the hoverbox either.

**Comment by sashaandjic** [comment]
screen shot 2017-08-15 at 1 19 08 pm

File: components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js [hunk 3]
```diff
@@ -267,6 +270,25 @@ define([
...
+                    raphaelElement.node.children[0].nodeName == "tspan") {
+                    raphaelElement.node.children[0].setAttribute("class","donothidehoverbox");
+                }
+            } else {
+                var _this = this;
+                raphaelElement.forEach(function(element) {
+                    _this.setDoNotHideHoverbox(element);
+                });
+            }
+        },
```
Comment (marta-) [inline components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js]: This function worries me a little. There are A LOT of elements with this class name in a pedigree (e.g. 480 in a family with 21 members). Additionally, some class attributes look weird, see for example:

image

Comment (allasm) [inline components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js]: re: screenshot: right, I see a problem (does not affect functionality, still repetitions and null-s better be fixed)

re: a lot of elements with this class name: well, this is what it has to be, is there an inherent problem with this? We just have a LOT of elements in each hoverbox (e.g. "alive & well" is composed of like 20 pieces, then 4 handles x a few pieces each, etc, etc), and a lot of hoverboxes.

Comment (marta-) [inline components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js]: Not a huge problem with so many elements with that class name, just wondering if there's a more efficient way, e.g. by grouping elements semantically. But this is definitely fine now (hence the approval of the PR).

Comment (allasm) [inline components/pedigree/resources/src/main/resources/pedigree/view/abstractHoverbox.js]: fixed problem #1 (see commit below), after removing class repetitions there are ~100 elements with this class name for a pedigree of 3 people

**Pull request merged**